text/0000-sandbox-environment.md

```diff
@@ -0,0 +1,235 @@
+
+By default all environment variables are available with their value taken from
+the process environment. However there are several command-line options to
+control this environment:
+- `--env-clear` - remove all process environment variables from the logical
+  environment, leaving it completely empty.
+- `--env-remove VARIABLE` - Remove a specific variable from the logical
+  environment. This is an exact match, and it does nothing if that variable is
+  not set.
+- `--env-pass VARIABLE` - Pass a variable from the process environment to the
+  logical one, even if it had previously been removed. This lets specific
+  variables to be allow-listed without having to explicitly set their value. The
+  variable is ignored if it is not set or not UTF-8 encoded.
+- `--env-set VARIABLE=VALUE` - Set a variable in the logical environment. This will
+  either create a new variable, or override an existing value.
+
+The options are processed in the order listed above (ie, clear, then remove,
+then pass, then set). Multiple `--env-set` options affecting the same variable are
+processed in command-line order, so later ones override earlier ones.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation of this RFC introduces the notion of:
+- the process environment which is inherited by the `rustc` process from it's
+  invoker, and
+- a logical environment which is accessed by the `env!`/`option_env!` macros
+
+The logical environment is initialized from the complete process environment,
+excluding only environment variables which are not UTF-8 encoded (name or
+value).
+
+Once initialized, the logical environment may be manipulated via the `--env-`
+command-line options described below.
+
+These environments are fundamentally key-value mappings, which is how they're
+represented within the session state - a map from `String` to `String`.
+
+## Processing of the options
+
+These options are processed in order:
+1. `--env-clear` - remove all variables from the logical environment.
+1. `--env-remove VAR` - remove a specific variable from the logical environment.
+   May be specified multiple times.
+1. `--env-pass VAR`- set a variable in the logical environment from the process
+   environment. Ignored if the variable is not set, or is not UTF-8 encoded.
+1. `--env-set VAR=VALUE` - multiple `--env-set` options affecting the same variable are
+   processed in command-line order, so later ones override earlier ones.
+
+`rustc` will only accept UTF-8 encoded command-line options, which affects all
+these options. This implies that all environment variables must have UTF-8
+encoded names and values.
+
+## Compile-time behaviour
+
+The `env!()` and `option_env!()` macros only inspect the logical environment
+with no reference to the process environment.
+
+Note that this can't affect other environment accesses. For example, if a
+procedural macro uses the `std::env::var()` function as part of its
```

Comment:
I think the design questions around proc macros deserve their own section, with a proper discussion of the alternatives. I think we need to consider at least all of:

 1. We do magic to modify the behavior of `env::var()` in proc macros
 2. We leave the behavior of `env::var()` as-is, but modify the behavior of `proc_macro::tracked_env` or any similar APIs that exist in the future
 3. What bjorn suggested [here](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Add.20option.20to.20pass.20environment.20variables.20compiler-team.23653/near/374982385)
 4. Something else?

Reply:
There are semantically two types of getenv at compile time:
1. intended to get the environment of the actual running rustc process, for something that's needed at runtime - eg getting `$PATH` to find an auxillary executable ("process environment")
2. environment logically substituted into the source being compiled - ie `env!()` etc. This is the only environment affected by this RFC ("logical environment")

A proc macro could be doing either, so rather than making `env::var()` behave differently we'd need a new API to enable 2 in proc macros, and an audit of proc macros which fetch from the environment to see which ones they should call.

> What bjorn suggested [here](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Add.20option.20to.20pass.20environment.20variables.20compiler-team.23653/near/374982385)

We should definitely discourage/block the use of `env::set_var` regardless of anything else.
Looks like [`proc_macro_tracked_env`](https://github.com/rust-lang/rust/issues/99515) is pretty much the API I'm thinking of here.

Another point to consider is the environment of build scripts, both consumed and produced. 
- for scripts doing probing (eg pkgconfig) then they definitely want the real process environment
- but I think `cargo:rustc-env=VAR=VALUE` is very likely to be intended to affect the logical environment